src/dune_rules/melange_rules.ml

```diff
@@ -16,43 +16,43 @@ let ocaml_flags sctx ~dir melange =
     with_vendored_flags ~ocaml_version flags
   | false -> flags
 
-let lib_output_dir ~target_dir ~lib_dir =
+let lib_output_dir ~target_dir lib =
+  let info = Lib.info lib in
+  let lib_dir =
+    match Lib.Local.of_lib lib with
+    | None ->
+      let package_name = Option.value_exn (Lib_info.package info) in
```

Comment:
Why is `value_exn` okay here?

Reply:
I don't know if it is. What would you suggest we do if it's `None`?

This brings up another design question:
- I think I'd rather use the library `public_name` as the `--bs-package-name ..` so that we could depend on multiple public libraries within the same package. The naming scheme for those is `alphanumerical-maybe-dashes.sublibrary`, which I think can be a valid folder name for the node.js module resolution algorithm

Reply:
> What would you suggest we do if it's None?

Work the way it worked before I guess? Right now this will not work for private libraries.

> I think I'd rather use the library public_name as the --bs-package-name .. so that we could depend on multiple public libraries within the same package. The naming scheme for those is alphanumerical-maybe-dashes.sublibrary, which I think can be a valid folder name for the node.js module resolution algorithm

Better ask @jchavarri about that.

Reply:
Why not? The tests are passing. This is in the branch where `Lib.Local.of_lib` is `None`. I thought that'd be the branch for external libraries.

Reply:
I see. But then you're distinguishing between public libraries that are installed and ones that are in the workspace.

I think you should distinguish between public (external or in-workspace) and private libraries instead.

Reply:
that works. What's a predicate I can use?

Reply:
We could also always use `Lib.name` now that we have the `node_modules` emission stuff.

Reply:
There's `Lib_info.Status.t`. In particular, when it's `Private (_, None)`

Reply:
> I think I'd rather use the library public_name as the --bs-package-name .. 
> Better ask @jchavarri about that.

I thought about this and I believe it should be fine. The only caveat is that in order to consume a public library from the same workspace, the library should be installed first, but I guess that's ok? Are workspace libraries "constantly installed" when running e.g. `dune build -w`?

Reply:
> In particular, when it's `Private (_, None)`

I am taking a stab at fixing this, but I am confused about this comment. If we want to distinguish between public and private, why does the second part of the payload (`None`, the package) matter?

I made all tests pass with something like this:

```
  let lib_dir =
    match Lib_info.status info with
    | Installed | Installed_private | Private (_, Some _) ->
      let package_name = Option.value_exn (Lib_info.package info) in
      let bctx, _ = Path.Build.extract_build_context_dir_exn target_dir in
      Path.Build.(
        relative
          ((relative bctx) "node_modules")
          (Package.Name.to_string package_name))
    | Private (_, None) | Public _ ->
      let lib = Lib.Local.of_lib_exn lib in
      let info = Lib.Local.info lib in
      Lib_info.src_dir info
```

but it does not make much sense to me.

Reply:
Public libraries have the following guarantee: their behavior is the same regardless of whether they're installed or are present in the current workspace.

The second part of the payload matters because there's also private libraries with the above guarantee. Those private libraries with the `package` field set.